tests/nn/data_parallel/test_sharded_ddp_pytorch_parity.py

```diff
@@ -64,132 +65,134 @@ def run_ddp_parity(
-
-        # The model should be synchronized in between the ranks at construction time, check that
-        check_same_model_params(sharded_ddp_model, ddp_model)
-
-        # Typical training loop, check that we get the exact same results as DDP
-        for i in range(NUMBER_BATCHS):
-            input_tensor = torch.rand((BATCH_SIZE, 2)).to(device)
-
-            def ddp_closure(input_tensor=input_tensor):
-                return closure(ddp_model, ddp_scaler, input_tensor, grad_accumulation)
-
-            def sharded_closure(input_tensor=input_tensor):
-                return closure(
-                    sharded_ddp_model,
-                    sharded_scaler,
-                    input_tensor,
-                    grad_accumulation,
-                    _manual_reduction=manual_reduction,
-                )
-
-            # Step/scale both
-            for _scaler, _closure, _optimizer in (
-                (ddp_scaler, ddp_closure, ddp_optimizer),
-                (sharded_scaler, sharded_closure, sharded_optimizer),
-            ):
-                if _scaler is not None:
-                    _ = _closure(input_tensor)
-                    _scaler.step(_optimizer)
-                    _scaler.update()
-
-            check_same_model_params(sharded_ddp_model, ddp_model, f"Rank: {rank} - Step {i} broke")
-
-            # Check that the two grad norm are equivalent
-            # NOTE: The grads can occasionally be NaNs, the scaler will skip the step in that case
-            # This is not ShardedDDP specific. If the grads are not NaN for DDP then they should also
-            # be valid for ShardedDDP
-            if clip_grad_norm:
-                total_norm = torch.nn.utils.clip_grad_norm_(ddp_model.parameters(), 0.3, norm_type=2.0)  # type: ignore
-                if not torch.isnan(total_norm):
-                    oss_total_norm = sharded_optimizer.clip_grad_norm(0.3, norm_type=2.0)
-                    assert torch.allclose(
-                        oss_total_norm, total_norm, atol=1e-2 if amp else 1e-8
-                    ), f"torch and fairscale should return the same grad norm\n {oss_total_norm} vs {total_norm}"
-                else:
-                    print(rank, "NaN grad norm in DDP", flush=True)
-
-            # Flip the trainability of the first parameter back and forth
-            if i == 0 and change_train_graph:
-                next(sharded_ddp_model.parameters()).requires_grad = not next(
-                    sharded_ddp_model.parameters()
-                ).requires_grad
-                next(ddp_model.parameters()).requires_grad = not next(ddp_model.parameters()).requires_grad
-                check_same_model_params(sharded_ddp_model, ddp_model, f"Rank: {rank} - Trainability refresh {i} broke")
+            model.reduce()
 
-    # Test all combinations: AMP, Accumulate, Change train graph, reduce buckets
-    manual_reductions = [False, True] if not grad_accumulation and not change_train_graph else [False]
```

Comment:
screw that, not foolproof, now filling in the full test matrix except for clearly "skip" labelled cases which don't make sense